redox/redox.go

```diff
@@ -306,116 +321,76 @@ func (h *Handler) RescheduleSubscriptionOrdersForPatient(ctx context.Context, pa
-		(procedureCodes.DisableSummaryReports != nil && code == *procedureCodes.DisableSummaryReports) {
-		return h.MatchPatientsForSubscriptionOrder(ctx, clinic, order, update)
-	} else if procedureCodes.CreateAccount != nil && code == *procedureCodes.CreateAccount {
-		return h.FindMatchingPatientsForAccountCreationOrder(ctx, clinic, order)
-	}
-
-	return nil, nil
-}
-
-func (h *Handler) MatchPatientsForSubscriptionOrder(ctx context.Context, clinic clinics.Clinic, order models.NewOrder, update *patients.SubscriptionUpdate) ([]*patients.Patient, error) {
-	criteria, err := GetPatientMatchingCriteriaFromNewOrder(order, clinic)
+	matchingPatients, err := h.findMatchingPatients(ctx, *clinic, matchOrder)
 	if err != nil {
 		return nil, err
 	}
-	if criteria == nil {
-		return nil, nil
-	}
 
-	clinicId := clinic.Id.Hex()
-	filter := patients.Filter{
-		ClinicId:  &clinicId,
-		Mrn:       &criteria.Mrn,
-		BirthDate: &criteria.DateOfBirth,
-	}
-
-	page := store.Pagination{
-		Offset: 0,
-		Limit:  100,
-	}
-
-	result, err := h.patients.List(ctx, &filter, page, nil)
-
-	if err == nil && result.TotalCount == 1 && result.Patients[0] != nil && update != nil {
-		// Update the subscription for matched patient only if single match was found
-		match := result.Patients[0]
-		if err := h.patients.UpdateEHRSubscription(ctx, match.ClinicId.Hex(), *match.UserId, *update); err != nil {
+	// Update the subscription for matched patient only if single match was found
+	if matchOrder.SubscriptionUpdate != nil && len(matchingPatients) == 1 {
+		match := matchingPatients[0]
+		err = h.patients.UpdateEHRSubscription(ctx, clinic.Id.Hex(), *match.UserId, *matchOrder.SubscriptionUpdate)
+		if err != nil {
 			return nil, err
 		}
 	}
 
-	return result.Patients, err
+	return &MatchResult{
+		Clinic:   *clinic,
+		Patients: matchingPatients,
+	}, nil
 }
 
-func (h *Handler) FindMatchingPatientsForAccountCreationOrder(ctx context.Context, clinic clinics.Clinic, order models.NewOrder) ([]*patients.Patient, error) {
-	criteria, err := GetPatientMatchingCriteriaFromNewOrder(order, clinic)
+func (h *Handler) findMatchingPatients(ctx context.Context, clinic clinics.Clinic, matchOrder MatchOrder) ([]*patients.Patient, error) {
+	criteria, err := GetPatientMatchingValuesFromNewOrder(matchOrder.Order, clinic)
 	if err != nil {
 		return nil, err
 	}
 	if criteria == nil {
 		return nil, nil
 	}
 
+	unique := map[string]struct{}{}
 	var matchingPatients []*patients.Patient
 
-	clinicId := clinic.Id.Hex()
-	page := store.Pagination{
-		Offset: 0,
-		Limit:  100,
-	}
-
-	filter := patients.Filter{
-		ClinicId: &clinicId,
-		Mrn:      &criteria.Mrn,
-	}
-	result, err := h.patients.List(ctx, &filter, page, nil)
+	filters, err := criteria.GetFilters(clinic.Id.Hex(), matchOrder.PatientAttributes)
 	if err != nil {
 		return nil, err
 	}
+	for _, filter := range filters {
+		// When matches are non-unique, the results are not useful to clients.
```

Comment:
The consumers of this function are generally looking for unique matches. During the normal course of operations, having a handful (say < 10) matches is conceivable, but having 100 matches, given the criteria, would strongly indicate an error or misconfiguration of a filter, or some other problem.

As such, and to make this comment clearer, I would suggest:
1. removing this comment
2. adding a GoDoc to the findMatchingPatients function like the following:

```
// findMatchingPatients based on a MatchOrder.
//
// The number of matching patients is limited to 100. It is expected that consumers of this
// function are searching for unique patient matches. It is normal for a small number (< 10)
// of patients to match, but 100 matches strongly indicates an unexpected filter behavior.
//
// Placing a limit of 100 prevents a misconfigured filter from needlessly returning all of a
// clinic's patients.
```